### DISCUSSION TITLE
Self-referential relationship with multiple intermediary tables

### DISCUSSION THREAD

**Discussion by dkbarn** [sqlalchemy/sqlalchemy#10958] [Usage Questions] [answered]
I have a somewhat advanced relationship that I'm trying to define, from a table to itself, but which involves passing through multiple tables to get there.

Here is a full example of what I'm trying to do. The relationship I want to define is `Entity.upstream_entities` which is effectively:

Entity -> Account -> AccountConnection -> Account -> Entity

```python
from __future__ import annotations

from sqlalchemy import create_engine, ForeignKey, select
from sqlalchemy.orm import DeclarativeBase, Mapped, mapped_column, relationship, selectinload, Session

class Base(DeclarativeBase):
    pass

class Entity(Base):
    __tablename__ = 'entity'
    id: Mapped[int] = mapped_column(primary_key=True)

    upstream_entities: Mapped[set[Entity]] = relationship(
        primaryjoin="Entity.id==Account.entity_id",
        secondary=(
            "join(Account, AccountConnection, Account.id==AccountConnection.dest_id)"
            ".join(Account, AccountConnection.source_id==Account.id)"
        ),
        secondaryjoin="Account.entity_id==Entity.id",
        viewonly=True,
        lazy="raise",
    )

class Account(Base):
    __tablename__ = 'account'
    id: Mapped[int] = mapped_column(primary_key=True)
    entity_id: Mapped[int] = mapped_column(ForeignKey("entity.id"))

class AccountConnection(Base):
    __tablename__ = 'account_connection'
    id: Mapped[int] = mapped_column(primary_key=True)
    source_id: Mapped[int] = mapped_column(ForeignKey("account.id"))
    dest_id: Mapped[int] = mapped_column(ForeignKey("account.id"))

engine = create_engine("sqlite://", echo=True)
Base.metadata.create_all(engine)
    
with Session(bind=engine) as session:
    new_entity = Entity()
    session.add(new_entity)

    result = session.scalars(
        select(Entity).options(selectinload(Entity.upstream_entities)).limit(1)
    ).first()
```

When I run the above program against SQLAlchemy 2.0.25, the SELECT statement it prints out is:
```
SELECT entity_1.id AS entity_1_id, entity.id AS entity_id
FROM entity AS entity_1 JOIN (
    account AS account_1
    JOIN account_connection AS account_connection_1 ON account_1.id = account_connection_1.dest_id
    JOIN account AS account_2 ON account_connection_1.source_id = account_1.id
) ON entity_1.id = account_1.entity_id
JOIN entity ON account_1.entity_id = entity.id
WHERE entity_1.id IN (?)
```
This isn't correct. Notice how the second account table is being joined on the wrong condition, referencing the first account table: `JOIN account AS account_2 ON account_connection_1.source_id = account_1.id`

The same thing happens when switching the engine from SQLite to a MySQL backend. And with MySQL it issues a warning too:
> sqlalchemy.exc.SAWarning: SELECT statement has a cartesian product between FROM element(s) "account_2" and FROM element "account_connection_1".  Apply join condition(s) between each element to resolve.

What am I doing wrong here?

**Comment by dkbarn** [comment]
I was able to get it to work by using an `aliased` reference, but that means I have to bind the relationship onto the Entity after the class definition, which I don't love (at least I'm assuming this is necessary, because I don't know of a way to use the `aliased` function inside a deferred string-based expression, so I've switched to non-string expressions, and due to the references to the Entity class, these can't be evaluated until after the Entity class definition is complete).

```python

class Entity(Base):
    __tablename__ = 'entity'
    id: Mapped[int] = mapped_column(primary_key=True)

upstream_account = aliased(Account)
Entity.upstream_entities = relationship(
    "Entity",
    primaryjoin=Entity.id==Account.entity_id,
    secondary=(
        join(Account, AccountConnection, Account.id==AccountConnection.dest_id)
        .join(upstream_account, AccountConnection.source_id==Account.id)
    ),
    secondaryjoin=upstream_account.entity_id==Entity.id,
    viewonly=True,
    lazy="raise",
)
```

A couple questions here:
1. Is there a way to make use of the `aliased(Account)` function inside a string-based expression?
2. If not, and I have to resort to doing this monkeypatching of the relationship onto the class after it's been defined, then is there any way for me to continue using the `Mapped[set[Entity]]` type annotation on the relationship?

**Reply by CaselIT** [reply]
Hi,

You can use a function to define the primary and secondary join. That can be a normal function too, it doesn't need to be a lambda

**Reply by jsormaz** [reply, marked as answer]
I've run into a similar problem. My solution is to bind the relationship using a mapper event. This is convenient because it lets you define it fully within the actual class, and if you use the class as a mixin it works there as well.

for you, it would look something like this:

```python
class Entity(Base):
    __tablename__ = 'entity'
    id: Mapped[int] = mapped_column(primary_key=True)

    @declared_attr
    def upstream_entities(cls):
        @event.listens_for(cls, "before_mapper_configured")
        def configure_upstream_entities(mapper, cls_):
            upstream_account = aliased(Account)
            cls.upstream_entities = relationship(
                "Entity",
                primaryjoin=Entity.id == Account.entity_id,
                secondary=(
                    join(
                        Account,
                        AccountConnection,
                        Account.id == AccountConnection.dest_id,
                    ).join(upstream_account, AccountConnection.source_id == Account.id)
                ),
                secondaryjoin=upstream_account.entity_id == Entity.id,
                viewonly=True,
                lazy="raise",
            )
        # dummy relationship for type hinting, it gets overwritten by the event listener
        return relationship("Entity", uselist=True, viewonly=True)
    
```

**Reply by dkbarn** [reply]
Clever! Thanks for sharing this approach, I never would have thought of it. I'll give it a try!
